src/hotspot/share/gc/g1/heapRegionManager.cpp

```diff
@@ -463,11 +463,12 @@ uint HeapRegionManager::find_highest_free(bool* expanded) {
   while (true) {
     HeapRegion *hr = _regions.get_by_index(curr);
     if (hr == NULL || !is_available(curr)) {
-      uint res = expand_at(curr, 1, NULL);
-      if (res == 1) {
-        *expanded = true;
-        return curr;
-      }
+      expand_exact(curr, 1, NULL);
```

Comment:
It seems that the original code handles the expansion failure case, but the new one doesn't. Better to have comments explaining why expansion should always success here.

It's not covered in this PR, but the documentation for `expand_exact` function is quite abstract, which could also be addressed in this PR, if you feel like it.

Reply:
Thanks for looking at this Albert. You are correct that this is a slight change of behavior, or at least it looks like that, but we should never fail this expand. That's why I added the `assert()` right after, to clearly state that after this call this region must be available. Are you thinking something like:
```suggestion
      // Region not available. Expand at this index to make it 
      // available. This call should never fail.
      expand_exact(curr, 1, NULL);
      assert(_regions.get_by_index(curr) != NULL && is_available(curr),
             "Region (%u) must be available after expand", curr);
```
I'm not sure it adds that much but if you like this better I can certainly add it. I will also improve the assertion to do:
```c++
assert(at(curr)->is_free(), "Region (%u) must be available and free after expand", curr);
```
The call to at() will do the other asserts and checking that the regions is free is a good improvement.

Reply:
> we should never fail this expand

I think I am missing some background knowledge. I assume `expand_exact()`, in generally, could fail, but for args, `curr, 1, NULL` here, it will not fail, because ... The "because" part should be in the comments right above the call.

As for the post-assert, `_regions.get_by_index(curr) != NULL && is_available(curr)`, I think they are already covered in `expand_exact`.

I like `at(curr)->is_free()` better, since the current function, `find_highest_free`, should return a region (by index) that's free.

Reply:
Looking at this again, I think the change in HeapRegionManager::find_highest_free() is not good.

While the call should never fail (it is done when allocating archive regions at VM startup when the heap is completely empty), the caller handles that - and does not allocate the archive regions at all.
Consider the use case where you start with a small heap, but a huge CDS archive (maybe even larger than the heap). This should still continue working without CDS usage.

Reply:
@albertnetymk, it depends on what you mean by fail, but not really, after a call to `expand_exact()` all regions in the given range are "available". There are no guarantee that all are free, but the HeapRegion-object has been created and the memory has been committed. In this case, since the regions was not available right before the call (and we only expand a single region), we know that it must now be both available and free. The new assert better shows the intent and maybe a comment like this will make it even clearer?
```suggestion
    // Found uncommitted and free region, expand to make it available for use.
    expand_exact(curr, 1, NULL);
```

Reply:
@tschatzl, not sure I follow the use case you describe and how using `expand_exact()` will differ from `expand_at()`. Since we're searching backwards in `HeapRegionManager::find_highest_free()` we know `expand_at()` can only expand precisely the requested region (all potentially higher regions must be in use already). If for some reason `expand_at()` would have expanded another region in the old code, this would be a bug, because we return the requested index.

Reply:
> it depends on what you mean by fail

Sth like `expand_exact(curr, (uint)-1, NULL)`. (Ofc, this is a contrived example.) After a closer look at the surrounding loop, I see it's actually iterating `curr` from `max_length()-1` to 0 with some early returns. (Seems to me a for-loop makes more sense.) Since `curr` always contains a value in this valid range, making it available (expanding one region) should always succeed.

> In this case, since the regions was not available right before the call (and we only expand a single region), we know that it must now be both available and free.

Maybe this is obvious to veterans, but making it explicit is more friendly for newcomers, IMO.

Reply:
> Sth like expand_exact(curr, (uint)-1, NULL). (Ofc, this is a contrived example.) 
True, with bad input `expand_exact()` will likely assert or crash. But as you've noticed, we make sure that the input is within bounds and valid. I agree that using a for-loop here is better, and I can file an enhancement for this. I would like this change to only address how we expand the found regions.

> Maybe this is obvious to veterans, but making it explicit is more friendly for newcomers, IMO.
I agree that making the code easier to understand is great, and to me using `expand_exact()` which can only expand this exact region should make it easier to follow. Even though the name doesn't say so `expand_at()` will expand any region starting at the given index, which to me is both wrong and very confusing for cases where you've found a specific region you want to make available.

Reply:
> I would like this change to only address how we expand the found regions.

Sure.

Thank you for the explanation. I think we have reached an agreement here; please update the pre-call comments and post-call assert as discussed above. Thank you again.

Reply:
Thanks! I also filed [JDK-8253420](https://bugs.openjdk.java.net/browse/JDK-8253420) for the `find_highest_free()` refactoring.

Reply:
Thank you.